ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/McpStreamableHttpConnectorExecutor.java

```diff
@@ -0,0 +1,180 @@
+import static org.opensearch.ml.common.CommonValue.MCP_TOOLS_FIELD;
+import static org.opensearch.ml.common.CommonValue.MCP_TOOL_DESCRIPTION_FIELD;
+import static org.opensearch.ml.common.CommonValue.MCP_TOOL_INPUT_SCHEMA_FIELD;
+import static org.opensearch.ml.common.CommonValue.MCP_TOOL_NAME_FIELD;
+import static org.opensearch.ml.common.CommonValue.TOOL_INPUT_SCHEMA_FIELD;
+import static org.opensearch.ml.common.connector.ConnectorProtocols.MCP_STREAMABLE_HTTP;
+
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.common.util.TokenBucket;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.ml.common.agent.MLToolSpec;
+import org.opensearch.ml.common.connector.Connector;
+import org.opensearch.ml.common.connector.McpStreamableHttpConnector;
+import org.opensearch.ml.common.exception.MLException;
+import org.opensearch.ml.common.input.MLInput;
+import org.opensearch.ml.common.model.MLGuard;
+import org.opensearch.ml.common.output.model.ModelTensors;
+import org.opensearch.ml.common.utils.StringUtils;
+import org.opensearch.ml.engine.annotation.ConnectorExecutor;
+import org.opensearch.ml.engine.tools.McpStreamableHttpTool;
+import org.opensearch.script.ScriptService;
+import org.opensearch.transport.client.Client;
+
+import com.google.gson.Gson;
+
+import io.modelcontextprotocol.client.McpClient;
+import io.modelcontextprotocol.client.McpSyncClient;
+import io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport;
+import io.modelcontextprotocol.spec.McpClientTransport;
+import io.modelcontextprotocol.spec.McpSchema;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@ConnectorExecutor(MCP_STREAMABLE_HTTP)
+public class McpStreamableHttpConnectorExecutor extends AbstractConnectorExecutor {
+
+    @Getter
+    private McpStreamableHttpConnector connector;
+    @Setter
+    @Getter
+    private TokenBucket rateLimiter;
+    @Setter
+    @Getter
+    private Map<String, TokenBucket> userRateLimiterMap;
+    @Setter
+    @Getter
+    private MLGuard mlGuard;
+
+    public McpStreamableHttpConnectorExecutor(Connector connector) {
+        super.initialize(connector);
+        this.connector = (McpStreamableHttpConnector) connector;
+    }
+
+    public List<MLToolSpec> getMcpToolSpecs() {
+        String mcpServerUrl = connector.getUrl();
+        String endpoint = connector.getParameters() != null && connector.getParameters().containsKey(ENDPOINT_FIELD)
+            ? connector.getParameters().get(ENDPOINT_FIELD)
+            : MCP_DEFAULT_STREAMABLE_HTTP_ENDPOINT;
+        if (mcpServerUrl == null) {
+            return Collections.emptyList();
```

Comment:
Is it possible `mcpServerUrl` is null ? I think we should validate when create MCP connector